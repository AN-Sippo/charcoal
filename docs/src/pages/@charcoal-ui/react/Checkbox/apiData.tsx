import { CheckboxProps } from '@charcoal-ui/react'
import { ComponentPropsWithRef } from 'react'
import { ApiTableData } from '../_components/ApiTable'

export const apiData: Omit<
  ApiTableData<CheckboxProps, HTMLInputElement>,
  keyof React.HTMLProps<HTMLInputElement>
> = {
  invalid: {
    description: '入力の不正化',
    default: 'false',
    type: 'boolean',
  },
}
