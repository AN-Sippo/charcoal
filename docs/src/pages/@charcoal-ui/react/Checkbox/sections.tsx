import { CheckboxProps } from '@charcoal-ui/react'
import { PreviewSection } from '../_components/Previews'

export const sections: PreviewSection<Omit<CheckboxProps, 'ref'>>[] = [
  {
    title: 'disabled',
    previewMetas: [
      {
        children: 'disabled',
        props: {
          children: undefined,
          disabled: true,
        },
      },
      {
        children: 'disabled',
        props: {
          children: undefined,
          disabled: true,
          checked: true,
        },
      },
    ],
  },
  {
    title: 'invalid',
    previewMetas: [
      {
        children: 'invalid',
        props: {
          children: undefined,
          invalid: true,
          checked: false,
        },
      },
      {
        children: 'invalid',
        props: {
          children: undefined,
          invalid: true,
          checked: true,
        },
      },
    ],
  },
]
