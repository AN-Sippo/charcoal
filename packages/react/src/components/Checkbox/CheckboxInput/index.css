.charcoal-checkbox-input {
  appearance: none;
  display: flex;
  cursor: pointer;
  margin: 0;
  width: 20px;
  height: 20px;
  border-radius: 4px;
  transition: 0.2s box-shadow, 0.2s background-color;
}

.charcoal-checkbox-input:disabled,
.charcoal-checkbox-input[readonly] {
  opacity: 0.32;
  cursor: default;
}

.charcoal-checkbox-input:checked {
  background-color: var(--charcoal-brand);
}

.charcoal-checkbox-input:checked::after {
  content: '';
  color: white;
  display: flex;
  margin: auto;
  width: 16px;
  height: 16px;
  background-image: url('data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22none%22%3E%3Cpath%20fill%3D%22%23FFFFFF%22%20d%3D%22M15.848%207.449a2%202%200%200%201%202.804%202.853l-8.664%208.514-4.402-4.402a2%202%200%201%201%202.828-2.828l1.598%201.598%205.836-5.735z%22%2F%3E%3C%2Fsvg%3E');
  background-repeat: no-repeat;
  background-size: contain;
  background-position: center;
}

.charcoal-checkbox-input:not(:checked) {
  border-width: 2px;
  border-style: solid;
  border-color: var(--charcoal-text4);
}

.charcoal-checkbox-input:not(:disabled):focus {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0, 150, 250, 0.32);
}
.charcoal-checkbox-input:not(:disabled):focus-visible {
  box-shadow: 0 0 0 4px rgba(0, 150, 250, 0.32);
}
.charcoal-checkbox-input:not(:disabled):focus:not(:focus-visible) {
  box-shadow: none;
}

.charcoal-checkbox-input:checked:not(:disabled):hover {
  background-color: var(--charcoal-brand-hover);
}
.charcoal-checkbox-input:checked:not(:disabled):active {
  background-color: var(--charcoal-brand-press);
}

.charcoal-checkbox-input[aria-invalid='true'],
.charcoal-checkbox-input[aria-invalid='true']:not(:disabled):focus {
  box-shadow: 0 0 0 4px rgba(255, 43, 0, 0.32);
}
