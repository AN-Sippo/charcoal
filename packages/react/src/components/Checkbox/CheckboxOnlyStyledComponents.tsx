import React, { forwardRef, memo } from 'react'
import styled, { css } from 'styled-components'
import { disabledSelector } from '@charcoal-ui/utils'

import Icon from '../Icon'

type CheckboxLabelProps =
  | {
      children: React.ReactNode
    }
  | {
      label: string
    }

export type CheckboxProps = CheckboxLabelProps & {
  readonly id?: string
  readonly name?: string

  readonly checked?: boolean
  readonly defaultChecked?: boolean
  readonly disabled?: boolean
  readonly readonly?: boolean

  readonly onClick?: () => void
  readonly onChange?: (isSelected: boolean) => void
  readonly onBlur?: () => void
  readonly onFocus?: () => void
}

const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
  function CheckboxInner(props, ref) {
    const isDisabled = (props.disabled ?? false) || (props.readonly ?? false)
    return (
      <InputRoot aria-disabled={isDisabled}>
        <CheckboxRoot>
          <CheckboxInput
            type="checkbox"
            aria-label={'label' in props ? props.label : undefined}
            ref={ref}
            checked={props.checked ?? false}
            disabled={props.disabled}
            readOnly={props.readonly}
          />
          <CheckboxInputOverlay
            aria-hidden={true}
            checked={props.checked ?? false}
          >
            <Icon name="24/Check" unsafeNonGuidelineScale={2 / 3} />
          </CheckboxInputOverlay>
        </CheckboxRoot>

        {'children' in props && <InputLabel>{props.children}</InputLabel>}
      </InputRoot>
    )
  }
)

export default memo(Checkbox)

const hiddenCss = css`
  visibility: hidden;
`

const InputRoot = styled.label`
  position: relative;
  display: flex;

  cursor: pointer;
  ${disabledSelector} {
    cursor: default;
    opacity: ${({ theme }) => theme.elementEffect.disabled.opacity};
  }

  gap: 4px;
`

const CheckboxRoot = styled.div`
  position: relative;
`

const CheckboxInput = styled.input`
  &[type='checkbox'] {
    appearance: none;
    display: block;
    cursor: pointer;
    margin: 0;
    width: 20px;
    height: 20px;

    transition: 0.2s box-shadow;

    &:checked {
      background: ${({ theme }) => theme.color.brand};
    }
    &:not(:checked) {
      border-width: 2px;
      border-style: solid;
      border-color: ${({ theme }) => theme.color.text4};
    }
    border-radius: 4px;

    :focus {
      box-shadow: 0 0 0 ${({ theme }) => theme.outline.default.weight}px
        ${({ theme }) => theme.outline.default.color};
    }
  }
`

const CheckboxInputOverlay = styled.div<{ checked?: boolean }>`
  position: absolute;
  top: -2px;
  left: -2px;
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: center;

  width: 24px;
  height: 24px;
  color: ${({ theme }) => theme.color.text5};

  ${({ checked }) => checked !== true && hiddenCss};
`

const InputLabel = styled.div`
  color: ${({ theme }) => theme.color.text2};

  font-size: 14px;
  /** checkbox の height が 20px なのでcheckbox と text が揃っているように見せるために行ボックスの高さを 20px にしている */
  line-height: 20px;
`
