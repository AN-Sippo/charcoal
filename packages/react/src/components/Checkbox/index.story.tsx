import { action } from '@storybook/addon-actions'
import Checkbox from '.'
import { Meta, StoryObj } from '@storybook/react'
import { useCallback, useState } from 'react'

export default {
  title: 'Checkbox',
  component: Checkbox,
} as Meta<typeof Checkbox>

export const Default: StoryObj<typeof Checkbox> = {
  args: {
    checked: false,
    defaultChecked: false,
    disabled: false,
    readonly: false,
    invalid: false,
  },
  render: (props) => {
    const [checked, setChecked] = useState(props.checked)
    const handleChange = useCallback((isSelected: boolean) => {
      setChecked(isSelected)
      action('change')(isSelected)
    }, [])

    return (
      <div>
        <Checkbox
          {...props}
          checked={checked ?? props.checked}
          name="labelled"
          label="label"
          onBlur={action('blur')}
          onClick={action('click')}
          onChange={handleChange}
          onFocus={action('focus')}
        >
          <span style={{ userSelect: 'none' }}>同意する</span>
        </Checkbox>
      </div>
    )
  },
}

export const Invalid: StoryObj<typeof Checkbox> = {
  args: {
    checked: false,
    defaultChecked: false,
    disabled: false,
    readonly: false,
  },
  render: (props) => {
    return (
      <Checkbox
        {...props}
        name="labelled"
        label="label"
        invalid
        onBlur={action('blur')}
        onClick={action('click')}
        onChange={action('change')}
        onFocus={action('focus')}
      >
        同意する
      </Checkbox>
    )
  },
}

export const Unlabelled: StoryObj<typeof Checkbox> = {
  args: {
    checked: false,
    defaultChecked: false,
    disabled: false,
    readonly: false,
    invalid: false,
  },
  render: (props) => {
    return (
      <div>
        <Checkbox
          {...props}
          name="unlabelled"
          label="label"
          onBlur={action('blur')}
          onClick={action('click')}
          onChange={action('change')}
          onFocus={action('focus')}
        />
      </div>
    )
  },
}
