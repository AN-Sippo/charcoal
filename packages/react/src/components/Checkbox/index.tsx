import './index.css'

import React, { forwardRef, memo } from 'react'
import { useId } from '@react-aria/utils'
import CheckboxInput, { CheckboxInputProps } from '../CheckboxInput'
import { useClassNames } from '../../_lib/useClassNames'

export type CheckboxProps = CheckboxInputProps & {
  children?: React.ReactNode
}

const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
  function CheckboxInner({ disabled, className, id, children, ...props }, ref) {
    const classNames = useClassNames('charcoal-checkbox__label', className)
    const htmlId = useId(id)
    const input = (
      <CheckboxInput {...props} id={htmlId} disabled={disabled} ref={ref} />
    )
    if (children === undefined) {
      return input
    }
    return (
      <label htmlFor={htmlId} aria-disabled={disabled} className={classNames}>
        {input}
        <div className="charcoal-checkbox__text">{children}</div>
      </label>
    )
  }
)

export default memo(Checkbox)
