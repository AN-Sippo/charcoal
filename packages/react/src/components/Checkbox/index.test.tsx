import 'jest-styled-components'

import React from 'react'
import Checkbox from '.'
import { render } from '../test_util'

describe('Checkbox', () => {
  test('<Checkbox />', () => {
    expect(render(<Checkbox label="test" />)).toMatchSnapshot()
  })
})
