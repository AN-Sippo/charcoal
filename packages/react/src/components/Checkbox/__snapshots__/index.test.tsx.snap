// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox <Checkbox /> 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  gap: 4px;
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled=false]) {
  cursor: default;
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled=false]) {
  opacity: 0.32;
}

.c1 {
  position: relative;
}

.c2[type='checkbox'] {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  cursor: pointer;
  margin: 0;
  width: 20px;
  height: 20px;
  border-radius: 4px;
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
  -webkit-transition: all 0.2s !important;
  transition: all 0.2s !important;
}

.c2[type='checkbox']:checked {
  background-color: var(--charcoal-brand);
  -webkit-transition: 0.2s background-color;
  transition: 0.2s background-color;
}

.c2[type='checkbox']:checked:hover:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:checked:hover[aria-disabled=false] {
  background-color: var(--charcoal-brand-hover);
}

.c2[type='checkbox']:checked:active:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:checked:active[aria-disabled=false] {
  background-color: var(--charcoal-brand-press);
}

.c2[type='checkbox']:not(:checked) {
  border-width: 2px;
  border-style: solid;
  border-color: #adadad;
}

.c2[type='checkbox']:not(:disabled):not([aria-disabled]):focus,
.c2[type='checkbox'][aria-disabled=false]:focus,
.c2[type='checkbox']:not(:disabled):not([aria-disabled]):active,
.c2[type='checkbox'][aria-disabled=false]:active {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c2[type='checkbox']:not(:disabled):not([aria-disabled]):focus:not(:focus-visible),
.c2[type='checkbox'][aria-disabled=false]:focus:not(:focus-visible),
.c2[type='checkbox']:not(:disabled):not([aria-disabled]):active:not(:focus-visible),
.c2[type='checkbox'][aria-disabled=false]:active:not(:focus-visible) {
  outline: none;
}

.c2[type='checkbox']:not(:disabled):not([aria-disabled]):focus-visible,
.c2[type='checkbox'][aria-disabled=false]:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c3 {
  position: absolute;
  top: -2px;
  left: -2px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  color: var(--charcoal-text5);
  visibility: hidden;
}

<label
  aria-disabled={false}
  className="c0"
>
  <div
    className="c1"
  >
    <input
      aria-label="test"
      checked={false}
      className="c2"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      onDragStart={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="checkbox"
    />
    <div
      aria-hidden={true}
      checked={false}
      className="c3"
    >
      <pixiv-icon
        name="24/Check"
        unsafe-non-guideline-scale={0.6666666666666666}
      />
    </div>
  </div>
</label>
`;
